### PR TITLE
Carriers repair fighters to 1.5*MinimumHull immediately if they are disabled

### PR DESCRIPTION
**Pull request by UnorderedSigh**
**Bugfix:** The bug is this: when a carrier docks a disabled fighter, it will not immediately repair the fighter as if it had been boarded. My fix is to have the carrier repair disabled fighters to 1.5*MinimumHull as if they had been boarded.

**Important: There are two competing solutions. Please discuss them here: https://github.com/endless-sky/endless-sky/issues/7862**

Corrects an issue discovered while debugging #7662, but it does not fix that issue. This is because, once the fighter has been disabled, it permanently retains the `DISABLED` event.

This is one of three PRs that, together, should fix issue #7862

## Fix Details
Before a carrier considers repairing its fighters, it will check to see if the fighter is disabled. If it is, it'll repair it as if it had just been boarded. This means:

```
if(ship.hull < minimumHull || ship.isDisabled)
    ...
    ship.hull = min(max(ship.hull, minimumHull * 1.5), static_cast<double>(maximumHull))
```

## Testing Done
This was tested with the problematic Stormhold mission described in #7662.

## Save File

[Friction Diction~3026-01-21  should work but does not.txt](https://github.com/endless-sky/endless-sky/files/10203568/Friction.Diction.3026-01-21.should.work.but.does.not.txt)

### CODE REVIEW

**Comment by Zitchas** [comment]
Just to clarify, is the bug that carriers will undisable fighters? Or that they wouldn't? Based on the title and description, I'm not sure whether it is describing the problem or the new behavior.

**Comment by UnorderedSigh** [comment]
> Just to clarify, is the bug that carriers will undisable fighters? Or that they wouldn't? Based on the title and description, I'm not sure whether it is describing the problem or the new behavior.

Sorry, I just rewrote the title and description. Does it make sense now?

**Comment by UnorderedSigh** [comment]
The reason this is needed is that the only way we can clear `DISABLED` event on a fighter that has docked in a carrier is if we know for certain that docked fighters are no longer disabled.

Maybe I should put that fix in here too, so the whole thing is complete. It won't make much sense without that.

**Comment by warp-core** [comment]
This seems like a reasonable thing to do. It doesn't make sense for a carrier to collect a disabled fighter/drone and not repair it.

**Comment by UnorderedSigh** [comment]
> This seems like a reasonable thing to do. It doesn't make sense for a carrier to collect a disabled fighter/drone and not repair it.

Should I have this PR add DOCKED and DEPLOYED ShipEvents, so we can clear a fighter's DISABLED event?

Or does that belong in a second PR?

File: source/Ship.cpp [hunk 1]
```diff
@@ -2211,9 +2211,22 @@ void Ship::DoGeneration()
 			for(const pair<double, Ship *> &it : carried)
 			{
 				Ship &ship = *it.second;
+				int minimumHull = ship.MinimumHull();
+				int maximumHull = -1;
```
Comment (RisingLeaf) [inline source/Ship.cpp:2215]: why not like this?
```suggestion
				int maximumHull = ship.attributes.Get("hull");
```

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

**Comment by Zitchas** [comment]
Yes, that is much clearer and makes sense. Thanks!

Definitely sounds like something we should be doing. I agree it doesn't make sense to pick up fighters and leave them disabled.

**Comment by RisingLeaf** [comment]
> > This seems like a reasonable thing to do. It doesn't make sense for a carrier to collect a disabled fighter/drone and not repair it.
> 
> Should I have this PR add DOCKED and DEPLOYED ShipEvents, so we can clear a fighter's DISABLED event?
> 
> Or does that belong in a second PR?

I would say a second PR

File: source/Ship.cpp (hunk 1 above)
Comment (UnorderedSigh) [reply source/Ship.cpp:2215]: That will trigger a dictionary lookup, once per docked ship per frame.

**Comment by UnorderedSigh** [comment]
> I would say a second PR

That's just as well, since I haven't found the places I need to modify to make that happen.

File: source/Ship.cpp (hunk 1 above)
Comment (RisingLeaf) [reply source/Ship.cpp:2215]: Int his case, remove the variable and just place a get function in both places, as the additional if and the memory allocation are unneccessary imo

Comment (RisingLeaf) [reply source/Ship.cpp:2215]: or add a comment to explain why

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

**Comment by warp-core** [comment]
> > This seems like a reasonable thing to do. It doesn't make sense for a carrier to collect a disabled fighter/drone and not repair it.
> 
> Should I have this PR add DOCKED and DEPLOYED ShipEvents, so we can clear a fighter's DISABLED event?
> 
> Or does that belong in a second PR?

Is that event usually cleared when a ship is assisted?
If so, then it probably ought to be cleared when a carried ship is repaired upon being collected by a carrier. Is it not possible to do that when repairing the ship?

File: source/Ship.cpp (hunk 1 above)
Comment (UnorderedSigh) [reply source/Ship.cpp]: It is cleaner now; I moved the boarding-style repair into an AssistedRepair() function. How do you like it now?

**Comment by UnorderedSigh** [comment]
>  If so, then it probably ought to be cleared when a carried ship is repaired upon being collected by a carrier. Is it not possible to do that when repairing the ship?

It is cleared by sending an ShipEvent::ASSIST to the NPC. Somehow, the carrier would have to send an ASSIST event for the Ships in its bays.

**Review by RisingLeaf — APPROVED** [review]

**Comment by UnorderedSigh** [comment]
For the carrier to `ASSIST` the fighter, Ship::Move would need to send event information for its fighters to Engine::MoveShip. We'd need a new class, or a map<shared_ptr<Ship>,int> or something like that to pass the information.

It is far simpler for the fighter to send a DOCK vs. DEPLOY event for itself.

**Comment by UnorderedSigh** [comment]
The other reason ASSIST can't be used is this in NPC.cpp:

```
// Certain events only count towards the NPC's status if originated by
// the player: scanning, boarding, assisting, capturing, or provoking.
if(!event.ActorGovernment() || !event.ActorGovernment()->IsPlayer())
   type &= ~(ShipEvent::SCAN_CARGO | ShipEvent::SCAN_OUTFITS | ShipEvent::ASSIST
               | ShipEvent::BOARD | ShipEvent::CAPTURE | ShipEvent::PROVOKE);
```

**Comment by UnorderedSigh** [comment]
My not-quite-working fix to clear the DISABLED flag is here: #7852 but I can't figure out why the new REPAIRED_IN_BAY event isn't making it to the npc.

File: source/Ship.cpp [hunk 2]
```diff
@@ -2211,6 +2211,9 @@ void Ship::DoGeneration()
 			for(const pair<double, Ship *> &it : carried)
 			{
 				Ship &ship = *it.second;
+				if(ship.hull < ship.MinimumHull() || ship.isDisabled)
```
Comment (Hurleveur) [inline source/Ship.cpp:2214]: Do we need to check both of these? I think that isDisabled checks the hull vs mininum hull too?

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Looks good to me,maybe there could be another place that could make use of this new repair function though?

**Comment by vitalchip** [comment]
Carriers shouldn't be docking disabled fighters and we shouldn't be looking at a work around to something which should not have happened in the first place.

We should make sure it doesn't continue happening. Carried ships should be assisted first, then they can dock.

**Comment by UnorderedSigh** [comment]
> We should make sure it doesn't continue happening. Disabled carried ships should be assisted first, then they can dock.

That approach is not as good because:

1. If you land while an enemy npc carried ship is disabled and has a parent, in a system where they cannot land, it will be docked in the parent. 
2. Doing it the way you suggest will not fix existing saves that have this problem.
3. That adds an unneeded level of complication, making it more likely that the code will break in the future. Boarding the ship to repair it, and then docking it, is two steps over two frames, leaving more chances for things to go wrong. Docking a disabled ship is, in effect, the same as boarding and then docking it, but it is done in one frame, so the code is less likely to break.

**Comment by UnorderedSigh** [comment]
Actually, the ideal solution is to implement both fixes:

1. A carrier should board its disable fighters.
2. If the disabled fighter ends up docked somehow, it should be automatically repaired.

This way, you get the pilot confusion on the npc ship when it repairs the fighters (same disadvantage the player flagship has) and you fix the various other situations this could happen.

**Comment by UnorderedSigh** [comment]
At long last, the second PR, which clears the DISABLED flag, is ready for review: #7852

**Comment by UnorderedSigh** [comment]
Closing due to merge of #7866

**Pull request closed without merging**